training/dataset.py

```diff
@@ -14,16 +19,88 @@ def __len__(self):
     def shuffle(self):
         np.random.shuffle(self.data)
 
-    def iterbatches(self, batch_size):
+    def iterbatches(self, batch_size: int) -> List[Any]:
         batch = 0
         for _ in range(self.num_batches(batch_size)):
             start = batch*batch_size
             end = min((batch+1)*batch_size, len(self.data))
             yield self.data[start:end]
             batch += 1
 
-    def split(self, frac):
+    def split(self, frac: float) -> Tuple["Dataset", "Dataset"]:
+        """Split dataset into two, using a fraction.
+
+
+        Parameters
+        ----------
+
+        frac: float
+            The proportion of the dataset to have to split on.
+
+        Returns
+        -------
+        tuple(Dataset, Dataset)
+            The split datasets, the first dataset containing the fractional percentage
+            and the second with the leftover.
+
+        """
+        if frac > 1.0 or frac < 0.0:
+            raise ValueError("frac must be between 0 and 1")
         split_idx = int(len(self.data)*frac)
         train_dataset = self.data[:split_idx]
         test_dataset = self.data[split_idx:]
         return Dataset(train_dataset), Dataset(test_dataset)
+
+    def random_split(self, frac: float) -> Tuple["Dataset", "Dataset"]:
+        """Shuffle data and split into two datasets, using a fraction.
+
+
+        Parameters
+        ----------
+
+        frac: float
+            The proportion of the dataset to have to split on.
+
+        Returns
+        -------
+        tuple(Dataset, Dataset)
+            The split datasets, the first dataset containing the fractional percentage
+            and the second with the leftover
+
+        """
+        inds = np.arange(len(self))
+        np.random.shuffle(inds)
+        split_idx = int(len(self.data)*frac)
+        return self.indice_split(inds[:split_idx], inds[split_idx:])
+
+    def indice_split(self, left: List[int], right: List[int]) -> Tuple["Dataset", "Dataset"]:
```

Comment:
index (singular) indices (plural) - for sake of my OCD please name this either index_split or indices_split